src/rt-app.c

```diff
@@ -1277,8 +1292,11 @@ int main(int argc, char* argv[])
 		shutdown(SIGTERM);
 	}
 
-	for (i = 0; i < nthreads; i++) {
-		pthread_join(threads[i], NULL);
+	joining = 1;
+	for (i = 0; i < running_threads; i++) {
```

Comment:
Mmm, maybe it's just Friday evening :-), but I'm now actually wondering if the pthread_join(s) at the end of main() is ever executed?!

main sleeps for duration and then syncronously calls `shutdown(SIGTERM)` which (if continue_running is still set) wakes everybody up, wait for them and then exits after deconfiguring ftrace (if it was enabled). Shutdown is also executed by async signals.

Reply:
I ran a lot of tests with no duration was set, and that's how I hit the SIGINT issue. If duration was set, then yeah the pthread_join() logic in main will not be called as shutdown will call exit(EXIT_SUCCESS) once its pthread_join() is done. 

This makes me wonder though if this force wakeup logic called unconditionally makes sense when shutdown() is called from main :-/

Reply:
> with no duration was set

I think this is currently very broken.
So, if no duration was set, main created threads and then waits after joining them.
Then a SIGINT comes, it wakes everybody up and then joins the threads _again_!
According to man pages this is undefined behaviour both if main and shutdown call join concurrently and for whoever comes last. :-/
Maybe we could simply change main to wait for any characted from the user and call shutdown when it receives one (leaving to shutdown the join part)?

Reply:
Right, sorry, didn't want to put more work on you. :-)
Maybe the refactoring should still come first though.
Lemme see if I can find some time to do that.

Reply:
No I am happy to help. But if it's the only thing that blocks this PR to be merged, then I'd appreciate getting it merged first.

I cooked this quick patch though

```
diff --git a/src/rt-app.c b/src/rt-app.c
index 661c7d7..cdea703 100644
--- a/src/rt-app.c
+++ b/src/rt-app.c
@@ -24,6 +24,7 @@ Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 
 #include <stdlib.h>
 #include <stdio.h>
+#include <stdbool.h>
 #include <string.h>
 #include <fcntl.h>
 #include <math.h>
@@ -545,17 +546,10 @@ int run(int ind,
 	return perf;
 }
 
-static void
-shutdown(int sig)
+static void wakeup_all_threads(void)
 {
 	int i;
 
-	if(!continue_running)
-		return;
-
-	/* notify threads, join them, then exit */
-	continue_running = 0;
-
 	/* Force wake up of all waiting threads */
 	for (i = 0; i <  opts.nresources; i++) {
 		if (opts.resources[i].type == rtapp_wait) {
@@ -565,17 +559,30 @@ shutdown(int sig)
 			pthread_cond_broadcast(&opts.resources[i].res.barrier.c_obj);
 		}
 	}
+}
+
+static void __shutdown(bool force_wake_threads)
+{
+	int i;
+
+	if(!continue_running)
+		return;
+
+	/* notify threads, join them, then exit */
+	continue_running = 0;
+
+	if (force_wake_threads)
+		wakeup_all_threads();
 
 	/*
-	 * If we try to interrupt pthread_join() at the end of main then we
-	 * might end up joining the same thread twice which is undefined
-	 * behaviour and can lead to segfaults.
+	 * Make sure pthread_join() is done once or we hit undefined behaviour
+	 * of multiple simulataneous calls to pthread_join().
 	 *
-	 * Only join if nowhere else started the join process already.
+	 * This is a problem because __shutdown() could be called
+	 * asynchronously from some signals and from main().
 	 */
 	if (joining)
 		return;
-
 	joining = 1;
 
 	/* wait up all waiting threads */
@@ -595,6 +602,12 @@ shutdown(int sig)
 	exit(EXIT_SUCCESS);
 }
 
+static void
+shutdown(int sig)
+{
+	__shutdown(true);
+}
+
 static void create_cpuset_str(cpuset_data_t *cpu_data)
 {
 	unsigned int cpu_count = CPU_COUNT_S(cpu_data->cpusetsize,
@@ -1289,22 +1302,10 @@ int main(int argc, char* argv[])
 		sleep(opts.duration);
 		if (opts.ftrace)
 			log_ftrace(ft_data.marker_fd, "main shutdown\n");
-		shutdown(SIGTERM);
-	}
-
-	joining = 1;
-	for (i = 0; i < running_threads; i++) {
-		int ret = pthread_join(threads[i], NULL);
-		if (ret)
-			perror("pthread_join() failed");
+		__shutdown(false);
 	}
 
-	if (opts.ftrace) {
-		log_ftrace(ft_data.marker_fd, "main ends\n");
-		close(ft_data.marker_fd);
-	}
-	exit(EXIT_SUCCESS);
-
+	__shutdown(false);
 
 exit_err:
 	exit(EXIT_FAILURE);
```

Reply:
I pushed this as a new patch "Refactor shutdown()"